File: lrucache.js
//Objective is to design an LRU with a 'get' and 'put' method. If the 
//cache cannot 'get' a certain key, it returns -1. If the cache goes
//over it's capacity, it pushes out it's least recently used key.


//Design of an LRUCache using a dictionary. It has an O(1) operation for
//getting and putting key and value pairs in.

class LRUCache {
    constructor(capacity) {
        this.cache = new Map()
        this.capacity = capacity
    }

    get(key) {
        if (!this.cache.has(val)) {
            return -1
        }
        let value = this.cache.get(key)
        this.cache.delete(key)
        this.cache.set(key, value)
        return this.cache.get(key)
    }

    put(key, value) {
        if (this.cache.has(key)) {
            this.cache.delete(key)
        }
        this.cache.set(key, value)

        if (this.cache.size > this.capacity) {
            this.cache.delete(this.cache.keys().next().value)
        }
    }
}

let newCache = new LRUCache(3)
newCache.put(1, 1)
newCache.put(2, 2)
newCache.put(3, 3)
newCache.put(4, 4)
return newCache


//O(n) alternative solution using a doubly linked list as a way to keep track of the order of the nodes

class Node {
    constructor(key, val) {
        this.key = key
        this.val = val
        this.prev = null
        this.next = null
    }
}

class LRUCacheWithLinkedList {
    constructor(capacity) {
        this.cache = new Map()
        this.capacity = capacity
        this.size = 0
        this.dummyHead = new Node(0, 0)
        this.dummyTail = new Node(0, 0)
        this.dummyTail.prev = this.dummyHead
        this.dummyHead.next = this.dummyTail 
    }
    
    remove(node) {
        node.next.prev = node.prev 
        node.prev.next = node.next
    }
    
    addToLast(node) {
        node.next = this.dummyTail
        node.prev = this.dummyTail.prev
        this.dummyTail.prev.next = node
        this.dummyTail.prev = node
    }

    get(key) {
        if (!this.cache.has(key)) {
            return -1
        }

        let target = this.cache.get(key)
        this.remove(target)
        this.addToLast(target)
        return target.val
    }

    put(key, value) {
        if (this.cache.has(key)) {
            let target = this.cache.get(key)
            target.val = value
            this.remove(target)
            this.addToLast(target)
        } else {
            if (this.size == this.capacity) {
                this.cache.delete(this.dummyHead.next.key)
                this.remove(this.dummyHead.next)
                this.size--
            }

            let newNode = new Node(key, value)
            this.cache.set(key, newNode)
            this.addToLast(newNode)
            this.size++
        }
    }
}

File: lrucache.java
//Java Solution

class LRUCache {
    
    class DLinkedNode {
        int key;
        int val;
        DLinkedNode prev;
        DLinkedNode next;
    }
    
    private void remove(DLinkedNode node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }
    
    private void addToLast(DLinkedNode node) {
        node.next = tail;
        node.prev = tail.prev;
        tail.prev.next = node;
        tail.prev = node;
    }
    
    private Map<Integer, DLinkedNode> cache = new HashMap<>();
    private int size;
    private int capacity;
    private DLinkedNode head, tail;

    public LRUCache(int capacity) {
        this.size = 0;
        this.capacity = capacity;
        head = new DLinkedNode();
        tail = new DLinkedNode();
        head.next = tail;
        tail.prev = head;
    }
    
    public int get(int key) {
        DLinkedNode node = cache.get(key);
        if (node == null) {
            return -1;
        }
        
        remove(node);
        addToLast(node);
        return node.val;
    }
    
    public void put(int key, int value) {
        DLinkedNode node = cache.get(key);
        
        if (node != null) {
            DLinkedNode node2 = cache.get(key);
            node2.val = value;
            remove(node2);
            addToLast(node2);
        } else {
            if (cache.size() == this.capacity) {
                cache.remove(this.head.next.key);
                remove(this.head.next);
                this.size--;
            }
            
            DLinkedNode newNode = new DLinkedNode();
            newNode.key = key;
            newNode.val = value;
            cache.put(key, newNode);
            addToLast(newNode);
            this.size++;
        }
    }
}